lib/util/source-code.js

```diff
@@ -106,86 +106,85 @@ function sortedMerge(tokens, comments) {
-     * To avoid this, the current implementation just matches newlines
-     * and uses match.index to get the correct line start indices.
-     */
-    while ((match = lineEndingPattern.exec(this.text))) {
-        this.lines.push(this.text.slice(this.lineStartIndices[this.lineStartIndices.length - 1], match.index));
-        this.lineStartIndices.push(match.index + match[0].length);
-    }
-    this.lines.push(this.text.slice(this.lineStartIndices[this.lineStartIndices.length - 1]));
+        /**
+         * The parsed AST for the source code.
+         * @type ASTNode
+         */
+        this.ast = ast;
 
-    this.tokensAndComments = sortedMerge(ast.tokens, ast.comments);
+        /**
+         * The source code split into lines according to ECMA-262 specification.
+         * This is done to avoid each rule needing to do so separately.
+         * @type string[]
+         */
+        this.lines = [];
+        this.lineStartIndices = [0];
 
-    // create token store methods
-    const tokenStore = new TokenStore(ast.tokens, ast.comments);
+        const lineEndingPattern = astUtils.createGlobalLinebreakMatcher();
+        let match;
 
-    for (const methodName of TokenStore.PUBLIC_METHODS) {
-        this[methodName] = tokenStore[methodName].bind(tokenStore);
-    }
+        /*
+         * Previously, this was implemented using a regex that
+         * matched a sequence of non-linebreak characters followed by a
+         * linebreak, then adding the lengths of the matches. However,
+         * this caused a catastrophic backtracking issue when the end
+         * of a file contained a large number of non-newline characters.
+         * To avoid this, the current implementation just matches newlines
+         * and uses match.index to get the correct line start indices.
+         */
+        while ((match = lineEndingPattern.exec(this.text))) {
+            this.lines.push(this.text.slice(this.lineStartIndices[this.lineStartIndices.length - 1], match.index));
+            this.lineStartIndices.push(match.index + match[0].length);
+        }
+        this.lines.push(this.text.slice(this.lineStartIndices[this.lineStartIndices.length - 1]));
 
-    // don't allow modification of this object
-    Object.freeze(this);
-    Object.freeze(this.lines);
-}
+        this.tokensAndComments = sortedMerge(ast.tokens, ast.comments);
 
-/**
- * Split the source code into multiple lines based on the line delimiters
- * @param {string} text Source code as a string
- * @returns {string[]} Array of source code lines
- * @public
- */
-SourceCode.splitLines = function(text) {
-    return text.split(astUtils.createGlobalLinebreakMatcher());
-};
+        // create token store methods
+        const tokenStore = new TokenStore(ast.tokens, ast.comments);
 
-SourceCode.prototype = {
-    constructor: SourceCode,
+        for (const methodName of TokenStore.PUBLIC_METHODS) {
+            this[methodName] = tokenStore[methodName].bind(tokenStore);
+        }
```

Comment:
If this `SourceCode` class extends `TokenStore` class, probably we can remove this copy of TokenStore's methods.